### PR TITLE
DRILL-8057: INFORMATION_SCHEMA filter push down is inefficient

### PR DESCRIPTION
**Pull request by jnturton**
# [DRILL-8057](https://issues.apache.org/jira/browse/DRILL-8057): INFORMATION_SCHEMA filter push down is inefficient

## Description

WHERE clauses in queries against INFORMATION_SCHEMA do not stop Drill from fetching a schema hierarchy from all enabled storage configs.  This results in abysmal performance when unresponsive data sources are enabled, as reported by users in the Apache Drill Slack channels.  This PR teaches info schema to prune irrelevant schema subtrees from the search tree and removes eager subschema registration by DrillSchemaFactory.

## Documentation
No user-visible change

## Testing
Existing info schema unit tests + one addition for IN operator
TestInfoSchemFilterPushDown#testLazySchemaRegistration to test that eager schema registration is not taking place.

### CODE REVIEW

File: exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestInfoSchema.java
Comment (vvysotskyi) [inline exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestInfoSchema.java:73]: Please add a new test instead of changing the existing one.

Comment (dzamo) [reply exec/java-exec/src/test/java/org/apache/drill/exec/sql/TestInfoSchema.java]: @vvysotskyi thanks, this was definitely not supposed to be included in what I pushed.  Reverted this test, another new one is already present.

**Comment by cgivre** [comment]
@dzamo Thanks for submitting this. Should we merge this as is, or do you think it is worth tackling some of the hackery in the StoragePluginRegistryImpl?

**Comment by dzamo** [comment]
@cgivre let's hold off I think.  While the change here would reasonably be mergeable (once checkstyle was satisfied), it does not bring any meaningful benefit if the earlier plugin scanning stays the way it is.  I'm still hopeful that something can be done

**Comment by jnturton** [comment]
@vvysotskyi one specific question that arises in this PR is that of closing schema objects.  Before this PR, `SchemaTreeProvider#close()` would recursively access the entire schema tree in order to call close on each schema object.  This resulted in a lookup of every storage plugin and then a call to `registerSchemas()` on it, even if the plugin did not participate in the query being run.  To avoid this I've made `SchemaTreeProvider#close()` a no-op but I'm not sure what the consequences of this might be.  IIRC the one implementation of `SchemaPlus#close()` I did check in Calcite was itself a no-op, making this change harmless.  But there could be other implementations of `SchemaPlus#close()`...

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaFilter.java
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaFilter.java:284]: I think it is possible when a user submits plan instead of query.

**Review by vvysotskyi — COMMENTED** [review]
@jnturton, I'm fine with the approach of using `createRootSchema` instead of `createFullRootSchema` to prevent building a full schema tree when it is called.

Regarding `addSchemasToCloseList`, I think it would be better to leave it as it is since in the future we might have some schema implementations that have logic in its close methods.

File: exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java
Comment (jnturton) [reply exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java:177]: @vvysotskyi I aimed here to work according to the existing design as I saw it.  DynamicRootSchema is a subtype of CalciteSchema which is a wrapper around a Schema stored in a member variable.  In the case of CalciteSchema it is the wrapped Schema that implements getSubSchema and getSubSchemaNames, not the CalciteSchema wrapper.  So I reasoned that for the case of DynamicRootSchema, the wrapped RootSchema should implement, indeed override, the mentioned methods.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/SchemaTreeProvider.java
Comment (jnturton) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/SchemaTreeProvider.java:178]: @vvysotskyi under this PR, the construction of the schema tree becomes lazy since createFullRootSchema is avoided.  The code at this line unfortunately undoes the benefits by eagerly constructing the entire schema tree only in order to be able to close schema objects.  I am not happy about skipping calls to `close()` on schemas, do you think I should look to add code to DynamicRootSchemas that returns only schemas that have already been constructed and initialised so that the code here has a suitable collection to close?  E.g. the `getSubSchemaNames()`, which returns every storage config name, gets replaced with something like `getInitialisedSubSchemas()`?

File: exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java:177]: I think the case of `DynamicRootSchema` is the special one, since it overrides the parent method `getImplicitSubSchema()` that uses the `Schema.getSubSchema()` method in the parent class. I think it was done intentionally, in the opposite case all the logic related to loading schemas simple would be placed in the `RootSchema`.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/SchemaTreeProvider.java
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/SchemaTreeProvider.java:178]: Yes, initially I had doubts about this change, but for now, I propose to delete this method. For the case when someone would need to add some special logic under it, this change might be reconsidered. But we should be sure that changes in this PR are covered by tests, so breaking this optimization would be easily detected.

File: exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java
Comment (jnturton) [reply exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java:177]: @vvysotskyi good catch, thanks!

**Comment by jnturton** [comment]
I've added @paul-rogers to the reviewers since he's familiar with the storage plugin registry and might spot gotchas that were missed in this less-than-obvious change set.

**Review by cgivre — APPROVED** [review]
I don't really think I can add much to what @vvysotskyi has already said, so LGTM +1.

File: exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java
Comment (jnturton) [reply exec/java-exec/src/main/java/org/apache/calcite/jdbc/DynamicRootSchema.java:188]: @cgivre here's the description on this method in AbstractSchema.  I believe it was added when the information_schema db was originally developed.
```
  /**
   * Reports whether to show items from this schema in INFORMATION_SCHEMA
   * tables.
   * (Controls ... TODO:  Doc.:  Mention what this typically controls or
   * affects.)
   * <p>
   *   This base implementation returns {@code true}.
   * </p>
   */
  public boolean showInInformationSchema() {
    return true;
  }
```

The override in this PR says "don't show the root schema in query results against information_schema".

**Review by vvysotskyi — APPROVED** [review]
@jnturton, thanks for making changes, +1

**Pull request merged**
